app/utils/JourneyChecker.scala

```diff
@@ -32,23 +35,89 @@ trait JourneyChecker extends IncomeSourcesUtils {
   implicit val ec: ExecutionContext
 
   val sessionService: SessionService
-  def withIncomeSourcesFSWithSessionCheck(journeyType: JourneyType)(codeBlock: => Future[Result])(implicit user: MtdItUser[_], hc: HeaderCarrier): Future[Result] = {
-    withIncomeSourcesFS {
-      journeyChecker(journeyType).flatMap {
-        case true => user.userType match {
-          case Some(Agent) => Future.successful(Redirect(controllers.incomeSources.add.routes.YouCannotGoBackErrorController.showAgent(journeyType.businessType)))
-          case _ => Future.successful(Redirect(controllers.incomeSources.add.routes.YouCannotGoBackErrorController.show(journeyType.businessType)))
-        }
-        case false => codeBlock
+
+  private lazy val isAgent: MtdItUser[_] => Boolean = (user: MtdItUser[_]) => user.userType match {
+    case Some(Agent) =>
+      true
+    case _ =>
+      false
+  }
+
+  private lazy val redirectUrl: (JourneyType, Option[String], Option[String], Option[String], Boolean) => MtdItUser[_] => Future[Result] =
+    (journeyType: JourneyType, reportingMethod: Option[String], taxYear: Option[String], incomeSourceId: Option[String], useDefault: Boolean) => user => {
+      (journeyType.operation, isAgent(user)) match {
+        case (Add, true) =>
+          Future.successful {
+            if (useDefault) Redirect(controllers.incomeSources.add.routes.ReportingMethodSetBackErrorController.showAgent(journeyType.businessType))
+            else Redirect(controllers.incomeSources.add.routes.IncomeSourceAddedBackErrorController.showAgent(journeyType.businessType))
+          }
+        case (Add, false) =>
+          Future.successful {
+            if (useDefault) Redirect(controllers.incomeSources.add.routes.ReportingMethodSetBackErrorController.show(journeyType.businessType))
+            else Redirect(controllers.incomeSources.add.routes.IncomeSourceAddedBackErrorController.show(journeyType.businessType))
+          }
+        case (Manage, _) =>
+          Future.successful {
+            Redirect(controllers.routes.HomeController.show()) //TODO: fix
```

Comment:
just wondering what need to be fixed here in TODO?

Reply:
Jordan and Ed will need to add the Redirects to their You Cannot Go Back pages